admin/ubuntu.py

```diff
@@ -0,0 +1,186 @@
+    unexpected_keys = set(
+        pclass_kwargs.keys()
+    ).difference(
+        set(pclass_fields.keys())
+    )
+    if unexpected_keys:
+        raise ValueError(
+            "Unexpected keyword arguments for '{}'. "
+            "Found: {}. "
+            "Expected: {}.".format(
+                pclass_type.__name__,
+                list(unexpected_keys),
+                pclass_fields.keys(),
+            )
+        )
+
+    fields_to_validate = {
+        key: field
+        for key, field
+        in pclass_fields.items()
+        if key in pclass_kwargs
+    }
+    filter_type = type(
+        "Filter_{}".format(pclass_type.__name__),
+        (PClass,),
+        fields_to_validate
+    )
+    return filter_type(**pclass_kwargs)
+
+
+def url_records(url, record_type):
+    headings = HEADINGS_BY_TYPE[record_type]
+    for line in urlopen(url).readlines():
+        yield record_type(
+            **dict(zip(
+                headings,
+                line.decode('utf-8').rstrip("\n").split("\t"))
+            )
+        )
+
+
+def filter_records(records, search_record):
+    for record in records:
+        if pclass_issubset(
+                subset=search_record,
+                superset=record
+        ):
+            yield record
+
+
+def latest(**kwargs):
+    [latest] = filter_records(
+        records=url_records(
+            url_mangle(URL_LATEST_TEMPLATE, kwargs),
+            UbuntuBase,
+        ),
+        search_record=filter_for_pclass(
+            pclass_type=UbuntuBase,
+            pclass_kwargs=dict(
+                (k, kwargs[k])
+                for k in ("ubuntu_name", "ubuntu_variant", "release_cycle")
+            )
+        )
+    )
+    return filter_records(
+        records=url_records(
+            url_mangle(URL_DETAILS_TEMPLATE, kwargs),
+            UbuntuDetail,
+        ),
+        search_record=filter_for_pclass(
+            pclass_type=UbuntuDetail,
+            pclass_kwargs=freeze(kwargs).set(
+                "release_date", latest.release_date,
+            )
+        )
+    )
+
+
+class AMISearchUbuntuOptions(Options):
+    """
+    Options.
+    """
+    optParameters = [
+        ['release-cycle', None, u"daily",
+         'One of `daily` or `release`.', unicode],
+        ['ubuntu-name', None, u"trusty",
+         'An Ubuntu release name.', unicode],
+    ]
+
+
+def ami_search_ubuntu_main(args, top_level, base_path):
+    options = AMISearchUbuntuOptions()
+    try:
+        options.parseOptions(args)
+    except UsageError as e:
+        sys.stderr.write("%s: %s\n" % (base_path.basename(), e))
+        raise SystemExit(1)
+
+    print json.dumps(
+        {
+            r.region: r.ami_id
+            for r in latest(
+                release_cycle=options["release-cycle"],
+                ubuntu_name=options["ubuntu-name"],
+                ubuntu_variant=u"server",
+                ec2_image_type=u'ebs',
+                architecture=u'amd64',
+                hypervisor=u'hvm',
```

Comment:
For `hypervisor` it's a choice between `hvm` or `paravirtualization`  and we're using the former in our installer images because it's now the recommended image type. More here:
 * https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html

And it's `amd64` vs `i386`. Centos use `x86_64` . Since we only ever install Flocker on 64bit machines it didn't seem like something we need to make configurable.